src/librustc_resolve/lib.rs

```diff
@@ -2521,8 +2570,9 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
                                                                   ValueNS,
                                                                   impl_item.span,
                                                 |n, s| MethodNotMemberOfTrait(n, s));
-
-                                            visit::walk_impl_item(this, impl_item);
+                                            this.with_tuple_struct_self_ctor_rib(self_type, |this| {
```

Comment:
Ideally, `Self` value "rib" should be added in all cases when `Self` type rib is added, and not only for methods in impls.

Reply:
Fixed.